ksql-cli/src/main/java/io/confluent/ksql/util/CmdLineUtil.java

```diff
@@ -0,0 +1,199 @@
+   * <li>{@code "'escaped '' quote'"} becomes {@code ["escaped ' quote"]}</li>
+   * </ul>
+   *
+   * @return the list of parts.
+   */
+  public static List<String> splitByUnquotedWhitespace(final String string) {
+    return new WhitespaceParser().parse(string);
+  }
+
+  /**
+   * Remove any matched, i.e. closed, single quotes from the supplied {@code string}.
+   *
+   * <p>Works in a similar way to a unix command line would.
+   *
+   * <p>A quote can be escaped using a second single quote, e.g. {@code "'escaped '' quote'"}.
+   *
+   * <p>e.g.
+   * <ul>
+   * <li>{@code "unquoted"} becomes {@code "unquoted"}</li>
+   * <li>{@code "'some quoted string'"} becomes {@code "some quoted string"}</li>
+   * <li>{@code "'quoted'connected'quoted'"} becomes {@code "quotedconnectedquoted"}</li>
+   * <li>{@code "'escaped '' quote'"} becomes {@code "escaped ' quote"}</li>
+   * </ul>
+   *
+   * @return the input with matched quotes removed.
+   */
+  public static String removeMatchedSingleQuotes(final String string) {
+    return new QuoteRemover().parse(string);
+  }
+
+  /**
+   * Removes any trailing semi colon(s).
+   *
+   * @param s the string to trim.
+   * @return the trimmed string
+   */
+  public static String trimTrailingSemiColon(final String s) {
+    int pos = s.length() - 1;
+    while (pos >= 0 && s.charAt(pos) == ';') {
+      pos--;
+    }
+
+    return pos < 0 ? "" : s.substring(0, pos + 1);
+  }
+
+  private static final class WhitespaceParser {
+
+    private boolean inQuotes;
+    private boolean inWhitespace;
+    private String input;
+    private Builder<String> output;
+    private final StringBuilder currentToken = new StringBuilder();
+
+    private List<String> parse(final String s) {
+      inQuotes = false;
+      inWhitespace = false;
+      input = s.trim();
+      currentToken.setLength(0);
+      output = ImmutableList.builder();
+
+      for (int pos = 0; pos != input.length(); ++pos) {
+        pos = processCharAt(pos);
+      }
+
+      final String remainder = currentToken.toString();
+      if (!remainder.isEmpty()) {
+        output.add(remainder);
+      }
+
+      return output.build();
+    }
+
+    private int processCharAt(final int pos) {
+      final char c = input.charAt(pos);
+      int returnPos = pos;
+
+      if (!Character.isWhitespace(c)) {
+        inWhitespace = false;
+      } else if (!inQuotes && !inWhitespace) {
+        inWhitespace = true;
+
+        output.add(currentToken.toString());
+        currentToken.setLength(0);
+      }
+
+      if (!inWhitespace) {
+        currentToken.append(c);
+      }
+
+      if (c == '\'') {
+        if (!inQuotes) {
+          inQuotes = true;
+        } else if (isEscapedChar(input, pos)) {
+          ++returnPos;
+        } else {
+          inQuotes = false;
+        }
+      }
+      return returnPos;
+    }
+  }
+
+  private static final class QuoteRemover {
```

Comment:
I think that this and `WhitespaceParser` are both sufficiently complex to justify dedicated classes (and corresponding unit tests). I still need to go through and review them, but I'll do that with a fresh pair of eyes tomorrow morning :D

Reply:
IMHO they are implementation details of the methods that use them, and they are tested extensively through those methods. 

I'm not saying they've not got bugs ;) . Just that we gain nothing by promoting these to top level classes and testing them independently of the only place they are intended to be used.

Reply:
`(''[^'\s]*)|'(?<esc>([^']|'')*)'|(\S+)` This (kinda crazy) Regex seems to do the job pretty well, though at this point I'm not sure if I prefer this "magic" over the original:

```
    String text = "''t0 t1 'quoted string' 'quoted '' string'";

    final String NOT_DOUBLE = "''[^'\\s]*";
    final String ESCAPED = "([^']|'')*";
    final String NOT_WHITESPACE = "\\S+";

    String regex = String.format("(%s)|'(?<esc>%s)'|(%s)", NOT_DOUBLE, ESCAPED, NOT_WHITESPACE);

    Matcher m = Pattern.compile(regex).matcher(text);
    while (m.find()) {
      System.out.println(ObjectUtils.defaultIfNull(m.group("esc"), m.group()).replace("''","'"));
    }
```

This outputs
```
't0
t1
quoted string
quoted ' string
```

Reply:
Interesting!

Does the above regex work for all the test cases? I didn't think it was possible to match quotes with regex in the presence of escaped quotes.

Reply:
I have not checked - after you push the current code maybe I'll play around with it a little bit, time permitting.